File: src/llm.ts
import Anthropic from "@anthropic-ai/sdk";
import { resolve } from "node:path";
import {
  buildFileContext,
  buildRuleSection,
  buildAgenticTask,
  STATELESS_SYSTEM_PROMPT,
} from "./prompt.js";
import { DEFAULTS } from "./config.js";
import {
  AuthStorage,
  createAgentSession,
  SessionManager,
  type AgentSessionEvent,
} from "@mariozechner/pi-coding-agent";
import { getModel } from "@mariozechner/pi-ai";
import {
  ContextHint,
  FileCheckRequest,
  Rule,
  RuleVerdict,
  Verdict,
} from "./schema.js";
import type { ProgressReporter } from "./reporter.js";
import { withTimeout, delay } from "./concurrency.js";
import { parseVerdicts } from "./verdict-parser.js";

const MAX_RETRIES = 3;
const RETRY_BASE_DELAY_MS = 1000;

const VERDICT_TOOL: Anthropic.Tool = {
  name: "submit_verdict",
  description: "Submit the rule evaluation verdict as structured data.",
  input_schema: {
    type: "object" as const,
    properties: {
      reasoning: {
        type: "string",
        description: "1-3 sentences explaining the verdict, referencing specific code.",
      },
      line_refs: {
        type: "array",
        items: { type: "integer" },
        description: "Absolute line numbers of violations. Empty array for pass.",
      },
      context_hint: {
        type: "object",
        description: "Required only when verdict is needs-more-context.",
        properties: {
          read_files: { type: "array", items: { type: "string" } },
          question: { type: "string" },
        },
        required: ["read_files", "question"],
      },
      confidence: {
        type: "number",
        description: "Certainty 0.0–1.0. Use < 0.7 when genuinely ambiguous.",
      },
      verdict: {
        type: "string",
        enum: ["pass", "fail", "needs-more-context"],
      },
    },
    required: ["reasoning", "line_refs", "confidence", "verdict"],
  },
};

function buildVerdictFromToolInput(
  input: Record<string, unknown>,
  request: FileCheckRequest,
  fromAgentic = false,
): RuleVerdict {
  const rule = request.rules[0]!;
  const rawVerdict = input["verdict"] as string;
  const verdict: Verdict = ["pass", "fail", "needs-more-context"].includes(rawVerdict)
    ? (rawVerdict as Verdict)
    : "fail";

  const rawHint = input["context_hint"];
  const contextHint =
    rawHint && typeof rawHint === "object"
      ? {
          read_files: Array.isArray((rawHint as Record<string, unknown>)["read_files"])
            ? ((rawHint as Record<string, unknown>)["read_files"] as string[])
            : [],
          question:
            typeof (rawHint as Record<string, unknown>)["question"] === "string"
              ? ((rawHint as Record<string, unknown>)["question"] as string)
              : "",
        }
      : null;

  return {
    rule_id: rule.id,
    verdict,
    rule_severity: rule.severity,
    confidence: typeof input["confidence"] === "number" ? (input["confidence"] as number) : 0.5,
    reasoning:
      typeof input["reasoning"] === "string"
        ? (input["reasoning"] as string).replace(/\n/g, " ").replace(/\r/g, "").trim()
        : "",
    line_refs: Array.isArray(input["line_refs"]) ? (input["line_refs"] as number[]) : [],
    context_hint: contextHint,
    from_agentic: fromAgentic,
  };
}

/**
 * Internal agentic options with resolved values.
 * Used only inside llm.ts — callers use AgenticConfig from runner.ts.
 */
export interface AgenticOpts {
  timeoutMs: number;
  allowBash: boolean;
  model: string;
  // NOTE: apiKey removed — read from process.env["ANTHROPIC_API_KEY"] directly
}

export async function runStateless(
  client: Anthropic,
  request: FileCheckRequest,
  model: string,
  timeoutMs: number,
  cacheSystemPrompt: boolean = false,
  cacheFileContext: boolean = false,
  trace: boolean = false,
  progress?: ProgressReporter,
): Promise<RuleVerdict[]> {
  const rule = request.rules[0]!;
  const fileContext = buildFileContext(request, rule);
  const ruleSection = buildRuleSection(rule);

  const system: Anthropic.TextBlockParam[] | string = cacheSystemPrompt
    ? [{ type: "text", text: STATELESS_SYSTEM_PROMPT, cache_control: { type: "ephemeral" } }]
    : STATELESS_SYSTEM_PROMPT;

  const userContent: Anthropic.TextBlockParam[] | string = cacheFileContext
    ? [
        { type: "text", text: fileContext, cache_control: { type: "ephemeral" } },
        { type: "text", text: ruleSection },
      ]
    : fileContext + ruleSection;

  if (trace) {
    const sysText = typeof system === "string" ? system : system.map((b) => b.text).join("");
    const userText = typeof userContent === "string" ? userContent : userContent.map((b) => b.text).join("");
    process.stderr.write(`\n[TRACE] ── stateless prompt for ${request.file_path} [${request.rules[0]?.id ?? "?"}] ──\n`);
    process.stderr.write(`[TRACE] SYSTEM:\n${sysText}\n`);
    process.stderr.write(`[TRACE] USER:\n${userText}\n`);
  }

  for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
    try {
      const response = await withTimeout(
        client.messages.create({
          model,
          max_tokens: DEFAULTS.maxStatelessTokens,
          system,
          messages: [{ role: "user", content: userContent }],
          tools: [VERDICT_TOOL],
          tool_choice: { type: "tool", name: "submit_verdict" },
        }),
        timeoutMs
      );
      if (trace) {
        const toolUseBlock = response.content.find((b) => b.type === "tool_use");
        process.stderr.write(
          `[TRACE] RESPONSE for ${request.file_path} [${request.rules[0]?.id ?? "?"}]:\n` +
            JSON.stringify(toolUseBlock?.type === "tool_use" ? toolUseBlock.input : response.content, null, 2) + "\n"
        );
      }
      const toolUse = response.content.find((b): b is Anthropic.ToolUseBlock => b.type === "tool_use");
      if (toolUse) {
        return [buildVerdictFromToolInput(toolUse.input as Record<string, unknown>, request)];
      }
      // Fallback: model returned text despite forced tool_choice (should not happen)
      const rawText = response.content
        .filter((b): b is Anthropic.TextBlock => b.type === "text")
        .map((b) => b.text)
        .join("");
      return parseVerdicts(rawText, request);
    } catch (err) {
      const e = err as Error & { status?: number };
      if (e.message === "timeout") {
        const msg = `Stateless pass timed out (attempt ${attempt + 1}/${MAX_RETRIES}) [${request.file_path}]`;
        progress ? progress.log(msg) : process.stderr.write(msg + "\n");
      } else if (e.status === 429) {
        const msg = `Rate limited (attempt ${attempt + 1}/${MAX_RETRIES}) [${request.file_path}]`;
        progress ? progress.log(msg) : process.stderr.write(msg + "\n");
      } else if (e.status !== undefined && e.status >= 500) {
        const msg = `API server error ${e.status} (attempt ${attempt + 1}/${MAX_RETRIES}) [${request.file_path}]`;
        progress ? progress.log(msg) : process.stderr.write(msg + "\n");
      } else {
        throw err;
      }
    }
    if (attempt < MAX_RETRIES - 1) await delay(RETRY_BASE_DELAY_MS * Math.pow(2, attempt));
  }

  return request.rules.map((r) => ({
    rule_id: r.id,
    verdict: "fail" as Verdict,
    rule_severity: r.severity,
    confidence: 0.0,
    reasoning: "LLM call failed",
    line_refs: [],
    context_hint: null,
    from_agentic: false,
  }));
}

export async function runAgenticEscalation(
  request: FileCheckRequest,
  escalatedRules: [Rule],
  hints: (ContextHint | null)[],
  opts: AgenticOpts,
  progress?: ProgressReporter,
  trace: boolean = false
): Promise<RuleVerdict[]> {
  const repoRoot = resolve(request.repo_root);

  const tools = opts.allowBash
    ? ["read", "grep", "find", "ls", "bash"]
    : ["read", "grep", "find", "ls"];

  const model = getModel("anthropic", opts.model as Parameters<typeof getModel>[1]);
  if (!model) throw new Error(`Unknown model: ${opts.model}`);

  const authStorage = AuthStorage.create();
  const apiKey = process.env["ANTHROPIC_API_KEY"];
  if (!apiKey) {
    throw new Error("ANTHROPIC_API_KEY environment variable is required for agentic escalation");
  }
  authStorage.setRuntimeApiKey("anthropic", apiKey);

  const { session } = await createAgentSession({
    cwd: repoRoot,
    tools,
    model,
    authStorage,
    sessionManager: SessionManager.inMemory(),
  });

  const textBuffer: string[] = [];
  session.subscribe((event: AgentSessionEvent) => {
    if (
      event.type === "message_update" &&
      event.assistantMessageEvent.type === "text_delta"
    ) {
      textBuffer.push(event.assistantMessageEvent.delta);
    }
  });

  const agenticTask = buildAgenticTask(request, escalatedRules[0], hints.filter((h): h is ContextHint => h !== null));
  if (trace) {
    process.stderr.write(`\n[TRACE] ── agentic task for ${request.file_path} ──\n`);
    process.stderr.write(`[TRACE] TASK:\n${agenticTask}\n`);
  }

  try {
    const taskPromise = session.prompt(agenticTask);
    const timeoutPromise = new Promise<never>((_, reject) =>
      setTimeout(() => reject(new Error("agentic_timeout")), opts.timeoutMs)
    );
    await Promise.race([taskPromise, timeoutPromise]);
  } catch (err) {
    const e = err as Error;
    await session.abort();
    if (e.message === "agentic_timeout") {
      console.error(
        `Agentic escalation timed out after ${opts.timeoutMs}ms for ${request.file_path}`
      );
    } else {
      console.error(`Agentic escalation error for ${request.file_path}: ${e.message}`);
    }
  } finally {
    session.dispose();
  }

  const rawAgenticText = textBuffer.join("");
  if (trace) {
    process.stderr.write(`[TRACE] AGENTIC RESPONSE for ${request.file_path}:\n${rawAgenticText}\n`);
  }
  const subRequest: FileCheckRequest = { ...request, rules: escalatedRules };
  let verdicts = parseVerdicts(rawAgenticText, subRequest);

  // If the agent's final message wasn't parseable JSON, normalize via a single tool-use call.
  if (verdicts[0]?.reasoning === "JSON parse error" || verdicts[0]?.reasoning === "Model returned unrecognised verdict") {
    try {
      const normClient = new Anthropic({ apiKey: process.env["ANTHROPIC_API_KEY"] ?? "" });
      const normResponse = await normClient.messages.create({
        model: opts.model,
        max_tokens: 1024,
        system: "Extract a structured rule evaluation verdict from the agent's analysis. The agent has already done the investigation.",
        messages: [{
          role: "user",
          content:
            `File: ${request.file_path}\nRule: ${escalatedRules[0].id} — ${escalatedRules[0].name}\n\n` +
            `Agent analysis:\n${rawAgenticText.slice(0, 8000)}\n\nExtract the verdict.`,
        }],
        tools: [VERDICT_TOOL],
        tool_choice: { type: "tool", name: "submit_verdict" },
      });
      const toolUse = normResponse.content.find((b): b is Anthropic.ToolUseBlock => b.type === "tool_use");
      if (toolUse) {
        verdicts = [buildVerdictFromToolInput(toolUse.input as Record<string, unknown>, subRequest, true)];
      }
    } catch {
      // normalization failed — keep the parse-error fallback verdict
    }
  }

  return verdicts.map((v) => ({
    ...v,
    verdict: v.verdict === "needs-more-context" ? ("fail" as Verdict) : v.verdict,
    from_agentic: true,
  }));
}

File: rust/src/llm.rs
//! Anthropic API client with retry logic
//!
//! Uses reqwest for HTTP, thiserror for typed errors that support retry classification.

use serde::{Deserialize, Serialize};
use std::time::Duration;
use thiserror::Error;

use async_trait::async_trait;

use crate::evaluator::{StatelessEvalOpts, StatelessEvaluator};
use crate::prompt::{build_tool_schema, build_user_prompt, SYSTEM_PROMPT};
use crate::schema::{ContextHint, Rule, RuleContext, RuleVerdict, Verdict};

const API_BASE_URL: &str = "https://api.anthropic.com";
const API_VERSION: &str = "2023-06-01";
pub(crate) const MAX_RETRIES: u32 = 3;
pub(crate) const RETRY_BASE_DELAY_MS: u64 = 1000;

/// LLM-specific errors with retry classification
#[derive(Debug, Error)]
pub enum LlmError {
    /// API rate limit exceeded (HTTP 429)
    #[error("rate limited")]
    RateLimit,

    /// Non-retryable server error with status code
    #[error("server error: {0}")]
    ServerError(u16),

    /// Request timed out
    #[error("timeout")]
    Timeout,

    /// Authentication failure (invalid or missing API key)
    #[error("auth error: {0}")]
    Auth(String),

    /// Network or HTTP request failure
    #[error("request error: {0}")]
    Request(String),

    /// Failed to parse the API response body
    #[error("failed to parse response: {0}")]
    Parse(String),

    /// All retry attempts exhausted without success
    #[error("retries exhausted")]
    Exhausted,
}

impl LlmError {
    /// Returns true if this error is worth retrying
    pub fn is_retryable(&self) -> bool {
        matches!(
            self,
            LlmError::RateLimit | LlmError::ServerError(_) | LlmError::Timeout
        )
    }
}

/// Anthropic API client
pub struct AnthropicClient {
    client: reqwest::Client,
    api_key: String,
    base_url: String,
}

impl AnthropicClient {
    /// Create a new client with the Anthropic production endpoint
    ///
    /// # Errors
    /// Returns an error if the HTTP client cannot be constructed.
    pub fn new(api_key: String) -> Result<Self, LlmError> {
        let client = reqwest::Client::builder()
            .timeout(Duration::from_secs(120))
            .pool_max_idle_per_host(20)
            .pool_idle_timeout(Duration::from_secs(90))
            .build()
            .map_err(|e| LlmError::Request(format!("failed to build HTTP client: {}", e)))?;

        Ok(Self {
            client,
            api_key,
            base_url: API_BASE_URL.to_string(),
        })
    }

    /// Evaluate a file against a single rule, returning one verdict
    ///
    /// Backward-compatible wrapper; prefer using `StatelessEvaluator` for new code.
    #[allow(dead_code)]
    #[allow(clippy::too_many_arguments)]
    pub async fn evaluate(
        &self,
        file_path: &str,
        diff: &str,
        content: Option<&str>,
        rule: &Rule,
        is_new_file: bool,
        model: &str,
        max_diff_chars: usize,
        max_content_chars: usize,
        timeout: Duration,
    ) -> Result<RuleVerdict, LlmError> {
        self.evaluate_internal(
            file_path,
            diff,
            content,
            rule,
            is_new_file,
            model,
            max_diff_chars,
            max_content_chars,
            timeout,
        )
        .await
    }

    /// Evaluate a file against a single rule, returning one verdict
    #[allow(clippy::too_many_arguments)]
    async fn evaluate_internal(
        &self,
        file_path: &str,
        diff: &str,
        content: Option<&str>,
        rule: &Rule,
        is_new_file: bool,
        model: &str,
        _max_diff_chars: usize,
        _max_content_chars: usize,
        timeout: Duration,
    ) -> Result<RuleVerdict, LlmError> {
        let user_prompt = build_user_prompt(file_path, diff, content, rule, is_new_file);

        let request = MessagesRequest {
            model,
            max_tokens: 2048,
            system: SYSTEM_PROMPT,
            messages: vec![Message {
                role: "user",
                content: &user_prompt,
            }],
            tools: vec![build_tool_schema()],
            tool_choice: ToolChoice {
                type_: "tool",
                name: Some("submit_verdict"),
                disable_parallel_tool_use: None,
            },
        };

        let response = match self.call_with_retry(&request, timeout).await {
            Ok(r) => r,
            Err(LlmError::Auth(msg)) => return Err(LlmError::Auth(msg)),
            Err(LlmError::Exhausted) => {
                return Ok(RuleVerdict {
                    rule_id: rule.id.clone(),
                    rule_name: rule.name.clone(),
                    verdict: Verdict::Fail,
                    confidence: 0.0,
                    reasoning: "LLM call failed after retries".to_string(),
                    severity: rule.severity,
                    line_refs: vec![],
                    line: None,
                    cached: false,
                    from_agentic: false,
                    context_hint: None,
                });
            }
            Err(e) => return Err(e),
        };

        self.parse_verdict(&response, rule)
    }

    async fn call_with_retry(
        &self,
        request: &MessagesRequest<'_>,
        timeout: Duration,
    ) -> Result<MessagesResponse, LlmError> {
        let mut last_error = LlmError::Exhausted;

        for attempt in 0..MAX_RETRIES {
            match self.call_once(request, timeout).await {
                Ok(response) => return Ok(response),
                Err(e) if e.is_retryable() && attempt < MAX_RETRIES - 1 => {
                    let delay = RETRY_BASE_DELAY_MS * 2u64.pow(attempt);
                    tokio::time::sleep(Duration::from_millis(delay)).await;
                    last_error = e;
                }
                Err(e) => return Err(e),
            }
        }

        Err(last_error)
    }

    async fn call_once(
        &self,
        request: &MessagesRequest<'_>,
        timeout: Duration,
    ) -> Result<MessagesResponse, LlmError> {
        let url = format!("{}/v1/messages", self.base_url);

        let response = tokio::time::timeout(
            timeout,
            self.client
                .post(&url)
                .header("x-api-key", &self.api_key)
                .header("anthropic-version", API_VERSION)
                .header("content-type", "application/json")
                .json(request)
                .send(),
        )
        .await
        .map_err(|_| LlmError::Timeout)?
        .map_err(|e| LlmError::Request(e.to_string()))?;

        let status = response.status().as_u16();

        match status {
            200 => {
                let body = response
                    .json::<MessagesResponse>()
                    .await
                    .map_err(|e| LlmError::Parse(e.to_string()))?;
                Ok(body)
            }
            401 | 403 => {
                let text = response.text().await.unwrap_or_default();
                Err(LlmError::Auth(text))
            }
            429 => Err(LlmError::RateLimit),
            500..=599 => Err(LlmError::ServerError(status)),
            _ => {
                let text = response.text().await.unwrap_or_default();
                Err(LlmError::Request(format!("HTTP {}: {}", status, text)))
            }
        }
    }

    fn parse_verdict(
        &self,
        response: &MessagesResponse,
        rule: &Rule,
    ) -> Result<RuleVerdict, LlmError> {
        for block in &response.content {
            if block.type_ == "tool_use" && block.name.as_deref() == Some("submit_verdict") {
                if let Some(input) = &block.input {
                    let verdict_str = input
                        .get("verdict")
                        .and_then(|v| v.as_str())
                        .unwrap_or("fail");

                    let mut verdict = match verdict_str {
                        "pass" => Verdict::Pass,
                        "needs-more-context" => Verdict::NeedsMoreContext,
                        _ => Verdict::Fail,
                    };

                    let confidence = input
                        .get("confidence")
                        .and_then(|v| v.as_f64())
                        .unwrap_or(0.5)
                        .clamp(0.0, 1.0);

                    let reasoning = input
                        .get("reasoning")
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string();

                    let reasoning = if verdict == Verdict::NeedsMoreContext
                        && rule.context == RuleContext::Stateless
                    {
                        verdict = Verdict::Fail;
                        format!(
                            "{} [collapsed from needs-more-context: stateless rule]",
                            reasoning.trim()
                        )
                    } else {
                        reasoning
                    };

                    let line_refs: Vec<u64> = input
                        .get("line_refs")
                        .and_then(|v| v.as_array())
                        .map(|arr| arr.iter().filter_map(|v| v.as_u64()).collect())
                        .unwrap_or_default();
                    let line_refs_u32: Vec<u32> = line_refs
                        .iter()
                        .filter_map(|&l| u32::try_from(l).ok())
                        .collect();
                    let line = line_refs_u32.first().copied();

                    let context_hint =
                        input
                            .get("context_hint")
                            .and_then(|v| v.as_object())
                            .map(|obj| ContextHint {
                                read_files: obj
                                    .get("read_files")
                                    .and_then(|v| v.as_array())
                                    .map(|arr| {
                                        arr.iter()
                                            .filter_map(|v| v.as_str().map(String::from))
                                            .collect()
                                    })
                                    .unwrap_or_default(),
                                question: obj
                                    .get("question")
                                    .and_then(|v| v.as_str())
                                    .unwrap_or("")
                                    .to_string(),
                            });

                    return Ok(RuleVerdict {
                        rule_id: rule.id.clone(),
                        rule_name: rule.name.clone(),
                        verdict,
                        confidence,
                        reasoning,
                        severity: rule.severity,
                        line_refs: line_refs_u32,
                        line,
                        cached: false,
                        from_agentic: false,
                        context_hint,
                    });
                }
            }
        }

        Ok(RuleVerdict {
            rule_id: rule.id.clone(),
            rule_name: rule.name.clone(),
            verdict: Verdict::Fail,
            confidence: 0.0,
            reasoning: "No verdict received from LLM".to_string(),
            severity: rule.severity,
            line_refs: vec![],
            line: None,
            cached: false,
            from_agentic: false,
            context_hint: None,
        })
    }
}

#[async_trait]
impl StatelessEvaluator for AnthropicClient {
    async fn evaluate(
        &self,
        file_path: &str,
        diff: &str,
        content: Option<&str>,
        rule: &Rule,
        is_new_file: bool,
        opts: &StatelessEvalOpts,
    ) -> Result<RuleVerdict, LlmError> {
        self.evaluate_internal(
            file_path,
            diff,
            content,
            rule,
            is_new_file,
            &opts.model,
            opts.max_diff_chars,
            opts.max_content_chars,
            opts.timeout,
        )
        .await
    }
}

#[derive(Debug, Serialize)]
struct MessagesRequest<'a> {
    model: &'a str,
    max_tokens: u32,
    system: &'a str,
    messages: Vec<Message<'a>>,
    tools: Vec<serde_json::Value>,
    tool_choice: ToolChoice<'a>,
}

#[derive(Debug, Serialize)]
struct Message<'a> {
    role: &'a str,
    content: &'a str,
}

#[derive(Debug, Serialize)]
struct ToolChoice<'a> {
    #[serde(rename = "type")]
    type_: &'a str,
    #[serde(skip_serializing_if = "Option::is_none")]
    name: Option<&'a str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    disable_parallel_tool_use: Option<bool>,
}

#[derive(Debug, Deserialize)]
struct MessagesResponse {
    content: Vec<ContentBlock>,
    #[allow(dead_code)]
    stop_reason: Option<String>,
}

#[derive(Debug, Deserialize)]
struct ContentBlock {
    #[serde(rename = "type")]
    type_: String,
    #[serde(default)]
    name: Option<String>,
    #[serde(default)]
    input: Option<serde_json::Value>,
    #[allow(dead_code)]
    #[serde(default)]
    text: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::schema::{RuleContext, Severity};

    #[test]
    fn test_llm_error_retryable() {
        assert!(LlmError::RateLimit.is_retryable());
        assert!(LlmError::ServerError(500).is_retryable());
        assert!(LlmError::Timeout.is_retryable());
        assert!(!LlmError::Auth("unauthorized".into()).is_retryable());
        assert!(!LlmError::Parse("bad json".into()).is_retryable());
        assert!(!LlmError::Request("connection refused".into()).is_retryable());
        assert!(!LlmError::Exhausted.is_retryable());
    }

    #[test]
    fn test_parse_verdict_missing_tool_use() {
        let client = AnthropicClient::new("test-key".to_string()).unwrap();

        let response = MessagesResponse {
            content: vec![],
            stop_reason: Some("end_turn".to_string()),
        };

        let rule = Rule {
            id: "rule-1".to_string(),
            name: "Rule 1".to_string(),
            prompt: "test".to_string(),
            severity: Severity::Error,
            enabled: true,
            context: Default::default(),
            glob_include: vec![],
            glob_exclude: vec![],
            examples: vec![],
            needs_more_context_when: String::new(),
        };

        let verdict = client.parse_verdict(&response, &rule).unwrap();
        assert_eq!(verdict.verdict, Verdict::Fail);
        assert!(verdict.reasoning.contains("No verdict"));
        assert_eq!(verdict.rule_id, "rule-1");
    }

    #[test]
    fn test_parse_verdict_tool_use() {
        let client = AnthropicClient::new("test-key".to_string()).unwrap();

        let input = serde_json::json!({
            "verdict": "pass",
            "confidence": 0.9,
            "reasoning": "looks good",
            "line_refs": []
        });

        let response = MessagesResponse {
            content: vec![ContentBlock {
                type_: "tool_use".to_string(),
                name: Some("submit_verdict".to_string()),
                input: Some(input),
                text: None,
            }],
            stop_reason: Some("tool_use".to_string()),
        };

        let rule = Rule {
            id: "rule-1".to_string(),
            name: "Rule 1".to_string(),
            prompt: "test".to_string(),
            severity: Severity::Warn,
            enabled: true,
            context: Default::default(),
            glob_include: vec![],
            glob_exclude: vec![],
            examples: vec![],
            needs_more_context_when: String::new(),
        };

        let verdict = client.parse_verdict(&response, &rule).unwrap();
        assert_eq!(verdict.verdict, Verdict::Pass);
        assert_eq!(verdict.confidence, 0.9);
        assert_eq!(verdict.rule_id, "rule-1");
    }

    #[test]
    fn test_nmc_collapse_stateless_rule() {
        let client = AnthropicClient::new("test-key".to_string()).unwrap();

        let input = serde_json::json!({
            "verdict": "needs-more-context",
            "confidence": 0.5,
            "reasoning": "Need to check imports",
            "line_refs": [],
            "context_hint": {
                "read_files": ["src/utils.rs"],
                "question": "What does utils export?"
            }
        });

        let response = MessagesResponse {
            content: vec![ContentBlock {
                type_: "tool_use".to_string(),
                name: Some("submit_verdict".to_string()),
                input: Some(input),
                text: None,
            }],
            stop_reason: Some("tool_use".to_string()),
        };

        let rule = Rule {
            id: "test-rule".to_string(),
            name: "Test Rule".to_string(),
            prompt: "test".to_string(),
            severity: Severity::Warn,
            enabled: true,
            context: RuleContext::Stateless,
            glob_include: vec![],
            glob_exclude: vec![],
            examples: vec![],
            needs_more_context_when: String::new(),
        };

        let verdict = client.parse_verdict(&response, &rule).unwrap();

        assert_eq!(verdict.verdict, Verdict::Fail);
        assert!(
            verdict
                .reasoning
                .contains("[collapsed from needs-more-context: stateless rule]"),
            "reasoning should contain collapse annotation: {}",
            verdict.reasoning
        );
        assert!(verdict.context_hint.is_some());
    }

    #[test]
    fn test_nmc_not_collapsed_agentic_rule() {
        let client = AnthropicClient::new("test-key".to_string()).unwrap();

        let input = serde_json::json!({
            "verdict": "needs-more-context",
            "confidence": 0.5,
            "reasoning": "Need to check imports",
            "line_refs": []
        });

        let response = MessagesResponse {
            content: vec![ContentBlock {
                type_: "tool_use".to_string(),
                name: Some("submit_verdict".to_string()),
                input: Some(input),
                text: None,
            }],
            stop_reason: Some("tool_use".to_string()),
        };

        let rule = Rule {
            id: "test-rule".to_string(),
            name: "Test Rule".to_string(),
            prompt: "test".to_string(),
            severity: Severity::Warn,
            enabled: true,
            context: RuleContext::Agentic,
            glob_include: vec![],
            glob_exclude: vec![],
            examples: vec![],
            needs_more_context_when: String::new(),
        };

        let verdict = client.parse_verdict(&response, &rule).unwrap();

        assert_eq!(verdict.verdict, Verdict::NeedsMoreContext);
        assert!(
            !verdict.reasoning.contains("collapsed"),
            "reasoning should not have collapse annotation for agentic rule"
        );
    }
}
